### PR TITLE
Postgres to ElasticSearch reindexer

### PR DESCRIPTION
**Pull request by chdorner**
This PR adds:

* A class for batch indexing all (or a big set of) annotations: `h.api.search.index.BatchIndexing`
* A class for batch removing deleted annotations from the search index: `h.api.search.index.BatchDeleting`
* A celery task for an easy way to run both at the same time in a worker: `hypothesis-celery call h.indexer.reindex`

I've added more detailed comments in the code, but all SQL queries, ElasticSearch queries and operations have been optimized as far as possible for causing the least amount of load on the respective Postgres/ElasticSearch servers, and for keeping the memory footprint of the Python process running these as low as possible.

I've tested it out locally and running both the indexing and deleting takes between 5 and 5:30 minutes on my local machine with approximately 350,000 annotations.

### CODE REVIEW

File: h/api/models/annotation.py [hunk 1]
```diff
@@ -79,11 +80,10 @@ class Annotation(Base, mixins.Timestamps):
...
 
-    document_uris = sa.orm.relationship('DocumentURI',
-                                        viewonly=True,
-                                        foreign_keys='DocumentURI.uri_normalized',
-                                        primaryjoin='DocumentURI.uri_normalized == Annotation.target_uri_normalized')
-    documents = association_proxy('document_uris', 'document')
+    document = sa.orm.relationship('Document',
+                                   secondary='join(DocumentURI, Document, DocumentURI.document_id == Document.id)',
+                                   primaryjoin='Annotation.target_uri_normalized == DocumentURI.uri_normalized',
+                                   uselist=False)
```
Comment (nickstenning) [inline h/api/models/annotation.py]: Much nicer 👍

File: h/api/models/document.py [hunk 2]
```diff
@@ -26,8 +26,12 @@ class Document(Base, mixins.Timestamps):
     #        dependency on the annotator-store is removed, as it clashes with
     #        making the Postgres and Elasticsearch interface of a Document
     #        object behave the same way.
-    document_uris = sa.orm.relationship('DocumentURI', backref='document')
-    meta = sa.orm.relationship('DocumentMeta', backref='document')
+    document_uris = sa.orm.relationship('DocumentURI',
+                                        backref='document',
+                                        order_by='DocumentURI.updated.desc()')
```
Comment (nickstenning) [inline h/api/models/document.py]: If we're sorting on these do we want an index on them, too?

File: h/api/models/document.py [hunk 3]
```diff
@@ -26,8 +26,12 @@ class Document(Base, mixins.Timestamps):
...
+    meta = sa.orm.relationship('DocumentMeta',
+                               backref='document',
+                               order_by='DocumentMeta.updated.desc()')
```
Comment (nickstenning) [inline h/api/models/document.py]: Ditto.

File: h/api/search/index.py [hunk 4]
```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
...
+                                             chunk_size=100,
+                                             raise_on_error=False,
+                                             expand_action_callback=self._present)
+        errored = set()
+        for ok, item in indexing:
+            if not ok:
+                errored.add(item['index']['_id'])
+        return errored
+
+    def _present(self, annotation):
```
Comment (nickstenning) [inline h/api/search/index.py]: I'd suggest calling this "prepare" or something a bit more generic -- I'm worried that we're falling slowly into the trap of assuming that we'll always use a presenter to turn an annotation object into a thing for the search index, and that definitely doesn't have to be true!

File: h/api/search/index.py [hunk 5]
```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
...
+        action = {'index': {'_index': self.es_client.index,
+                            '_type': self.es_client.t.annotation,
+                            '_id': annotation.id}}
+        data = presenters.AnnotationJSONPresenter(self.request, annotation).asdict()
```
Comment (nickstenning) [inline h/api/search/index.py]: For similar reasons, perhaps add a FIXME/TODO here to note that the `AnnotationJSONPresenter` is by no means the right way to prepare a document for indexing from an annotation, but it serves in the interim to replicate the structure of the old index.

File: h/api/search/index.py [hunk 6]
```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
...
+        # This is using a windowed query for loading all annotations in batches.
+        # It is the most performant way of loading a big set of records from
+        # the database while still supporting eagerloading of associated
+        # document data.
+
+        updated = self.session.query(models.Annotation.updated). \
+                execution_options(stream_results=True). \
+                order_by(models.Annotation.updated.asc()).all()
+
+        windows = [updated[x:x+chunksize] for x in range(0, len(updated), chunksize)]
```
Comment (nickstenning) [inline h/api/search/index.py]: It doesn't look to me like there's a reason you'd need to store all of the intermediate values here, is there? Couldn't you store something like:

```python
windows = [(updated[x], updated[x+chunksize-1]) for x in range(0, len(updated), chunksize)]
```

If you were feeling really nice, I'd probably do something like:

```python
class Window(namedtuple('Window', ['start', 'end'])):
    pass
```

so you could then do

```python
windows = [Window(start=updated[x], end=updated[x+chunksize-1]) 
           for x in range(0, len(updated), chunksize)]
...

for window in windows:
    batch = basequery.filter(models.Annotation.updated.between(window.start, window.end)).all()
    ...
```

Comment (nickstenning) [inline h/api/search/index.py]: Probably also want to use `xrange` here given the size of `len(updated)` -- you'll need to add something to `_compat` for this.

File: h/api/search/index.py [hunk 7]
```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
...
+            filter(models.Annotation.id.in_(annotation_ids)).all()
+
+        for a in annotations:
+            yield a
+
+    def _eager_loaded_query(self):
+        return self.session.query(models.Annotation).options(
+            subqueryload(models.Annotation.document).subqueryload(models.Document.document_uris),
+            subqueryload(models.Annotation.document).subqueryload(models.Document.meta)
+        )
```
Comment (nickstenning) [inline h/api/search/index.py]: SQLAlchemy ^*&@ yeah! 🤘

File: h/api/search/index.py [hunk 8]
```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
...
+        return errored
+
+    def _present(self, id_):
+        action = {'delete': {'_index': self.es_client.index,
+                             '_type': self.es_client.t.annotation,
+                             '_id': id_}}
+        return (action, None)
+
+    def _pg_ids(self):
+        # This is using a Postgres cursor for better query performance.
```
Comment (nickstenning) [inline h/api/search/index.py]: If this works for now, that's fine, but in general we can't keep building a set of all the ids in the database. Presumably we could return these in batches and do a `difference_update` for each batch?

File: h/api/search/test/index_test.py [hunk 9]
```diff
@@ -101,6 +112,245 @@ def log(self, patch):
...
+
+    @pytest.fixture
+    def index(self, patch):
+        return patch('h.api.search.index.BatchIndexer.index')
+
+    @pytest.fixture
+    def streaming_bulk(self, patch):
+        return patch('h.api.search.index.es_helpers.streaming_bulk')
+
+    @pytest.fixture
```
Comment (nickstenning) [inline h/api/search/test/index_test.py]: Doesn't look like you're using it as a fixture any more so this should probably go.

File: h/indexer.py [hunk 10]
```diff
@@ -1,17 +1,26 @@
...
 
 __all__ = (
     'add_annotation',
     'delete_annotation',
+    'reindex',
 )
 
 
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
```
Comment (nickstenning) [inline h/indexer.py]: This should probably go?

**Comment by nickstenning** [comment]
Looking pretty good to me except for the few issues I've raised.

File: h/api/models/document.py (hunk 2 above)
Comment (chdorner) [inline h/api/models/document.py]: Added an index to both DocumentMeta/URI `updated` column.

When checking the query with `EXPLAIN ANALYZE ...` I saw that it isn't using an index for filtering the `document_id` column. Turns out Postgres is [not automatically adding indices for foreign keys](https://hyp.is/AVSqPuhUFtPwhMO7Dnnm/www.postgresql.org/docs/9.4/static/ddl-constraints.html). There is a migration for that as well now.

File: h/api/models/document.py (hunk 3 above)
Comment (chdorner) [inline h/api/models/document.py]: Ditto from above.

File: h/api/search/index.py (hunk 4 above)
Comment (chdorner) [inline h/api/search/index.py]: Good point. Fixed.

File: h/api/search/index.py (hunk 5 above)
Comment (chdorner) [inline h/api/search/index.py]: Done.

File: h/api/search/index.py (hunk 6 above)
Comment (chdorner) [inline h/api/search/index.py]: Done, done, and done.

File: h/api/search/test/index_test.py (hunk 9 above)
Comment (chdorner) [inline h/api/search/test/index_test.py]: Done

File: h/indexer.py (hunk 10 above)
Comment (chdorner) [inline h/indexer.py]: It's gone.

File: h/api/search/index.py (hunk 8 above)
Comment (chdorner) [inline h/api/search/index.py]: It is now returning a SQLAlchemy query and we use the `batch_iter` function in `deleted_annotation_ids`, very similar to the way we load ElasticSearch annotation ids. :+1:

**Comment by chdorner** [comment]
I've addressed all comments. It is now even more query / memory optimized!

**Comment by nickstenning** [comment]
LGTM.

**Pull request merged**